src/screens/MessageDetails.js

```diff
@@ -45,27 +45,34 @@ import { getIdentityFromSender } from '../util/identitiesUtils';
 
 export default class MessageDetails extends React.PureComponent {
 	async onSignMessage(scannerStore, accountsStore, sender) {
-		try {
-			if (sender.isLegacy) {
-				return this.props.navigation.navigate('AccountUnlockAndSign', {
-					next: 'SignedMessage'
-				});
-			}
-			const senderIdentity = getIdentityFromSender(
-				sender,
-				accountsStore.state.identities
-			);
-			const seedPhrase = await unlockSeedPhrase(
-				this.props.navigation,
-				senderIdentity
-			);
-			await scannerStore.signDataWithSeedPhrase(
-				seedPhrase,
-				NETWORK_LIST[sender.networkKey].protocol
-			);
+		if (
+			sender.biometricEnabled &&
+			(await scannerStore.signDataBiometric(sender.isLegacy))
+		) {
 			return navigateToSignedMessage(this.props.navigation);
-		} catch (e) {
-			scannerStore.setErrorMsg(e.message);
+		} else {
+			try {
+				if (sender.isLegacy) {
+					return this.props.navigation.navigate('AccountUnlockAndSign', {
+						next: 'SignedMessage'
+					});
+				}
+				const senderIdentity = getIdentityFromSender(
+					sender,
+					accountsStore.state.identities
+				);
+				const seedPhrase = await unlockSeedPhrase(
+					this.props.navigation,
+					senderIdentity
+				);
+				await scannerStore.signDataWithSeedPhrase(
+					seedPhrase,
+					NETWORK_LIST[sender.networkKey].protocol
+				);
+				return navigateToSignedMessage(this.props.navigation);
+			} catch (e) {
+				scannerStore.setErrorMsg(e.message);
+			}
```

Comment:
```suggestion
	async onSignMessage(scannerStore, accountsStore, sender) {
		try {
			if (sender.biometricEnabled) {
				await scannerStore.signDataBiometric(sender.isLegacy);
				return navigateToSignedMessage(this.props.navigation);
			}
			if (sender.isLegacy) {
				return this.props.navigation.navigate('AccountUnlockAndSign', {
					next: 'SignedMessage'
				});
			}
			const senderIdentity = getIdentityFromSender(
				sender,
				accountsStore.state.identities
			);
			const seedPhrase = await unlockSeedPhrase(
				this.props.navigation,
				senderIdentity
			);
			await scannerStore.signDataWithSeedPhrase(
				seedPhrase,
				NETWORK_LIST[sender.networkKey].protocol
			);
			return navigateToSignedMessage(this.props.navigation);
		} catch (e) {
			scannerStore.setErrorMsg(e.message);
		}
	}
```

At the same time, now there is no error handling if `signDataBiometric` returns false.

It is better to wrap `signDataBiometric` (in `ScannerStore.js`) into a promise, and throw error with like `e.message = "Biometric data not correct, please try again"`. So that the error could be catched by the block here and displayed by `scannerStore.setErrorMsg(e.message);`

Reply:
The only likely biometric error here will be an invalidated fingerprint, in which case it should fallback to non-biometric pin entry so the user can still sign.  The boolean return is the simplest way of doing that.  But I agree it should tell the user something.

Reply:
I see you have already wrapped them into a promise.

The setErrorMsg I think it only display a warning message on the screen, in the fingerprint authentication case, we could show, `failed, please try again`. if so then they could be grouped together into the `try` block, which may easy the flow. 

But I am not sure on the fingerprint authentication case, maybe that would be annoying since the sensor is checking all the time (e.g. 10 times per second). Since I have not try the feature on real device.